tests/integ/test_data_source_api.py

```diff
@@ -169,32 +169,34 @@ def test_dbapi_retry(session):
     IS_WINDOWS,
     reason="sqlite3 file can not be shared across processes on windows",
 )
-def test_parallel(session):
+@pytest.mark.parametrize("upper_bound, expected_upload_cnt", [(5, 3), (100, 1)])
```

Comment:
we should also include a test of behavior when the target table is completely empty/ all data are null

Reply:
test_empty_table